fence_gce: logging and non required parameters

helen-fornazier opened the pull request:
The following changes were made provide some of the features supported by the following script:  https://storage.googleapis.com/sapdeploy/pacemaker-gcp/gcpstonith

fence_gce: add support for stackdriver logging
Add --logging option to enable sending logs to google stackdriver

fence_gce: set project and zone as not required

Try to retrieve the GCE project if the script is being executed inside a
GCE machine if --project is not provided.
Try to retrieve the zone automatically from GCE if --zone is not
provided.

oalbrigt commented:
You should use the built-in logging mechanisms in the fencing library.

helen-fornazier commented:
right, I'll do a wrapper then, because I need to send the log to the GCE stackdriver

oalbrigt commented:
Yeah. It supports -v to show logs on console, and -D to log to file. Btw. You'll have to run make xml-upload" and add the updated metadata to the PR as well to avoid CI failing.

helen-fornazier commented:
Ah yes, sorry about that, I'll update the PR

oalbrigt commented:
No worries.

helen-fornazier commented:
@oalbrigt I updated the PR, could you take a look please?

oalbrigt commented:
Could you add it to the default logger, so we can do all the logging via logging.*()?
https://github.com/ClusterLabs/fence-agents/blob/master/lib/fencing.py.py#L693

helen-fornazier commented:
I am not sure I understand what you mean, I am using the default logger https://github.com/ClusterLabs/fence-agents/pull/211/commits/6b7fef09f556613707edad20295ce59fc607afd0#diff-4a0f3ce2f5196c812280cc52f7aa42b6R25

Do you mean to add a handler instead? To wrap the default behavior?

oalbrigt commented:
Yeah.

helen-fornazier commented:
@oalbrigt I updated the PR, could you please check now?
I also added another commit implementing power cycle, let me know if you prefer a different PR for that.

Thanks

helen-fornazier commented:
@oalbrigt fyi, I created a child Logger to avoid every message to be sent to the new handler, but as it is a child logger, it will call the default parent logging.* as well.

oalbrigt commented:
This is done by removing the : in the getopt line: "getopt" : "",

helen-fornazier commented:
> This is done by removing the : in the getopt line: "getopt" : "",

This works great, thanks

> I think you should add it to the main logger, as it makes sense to log these to where users look for their logs in general too (without the -v parameter only WARNING or higher are logged, so it wont "spam" the logs).

The problem is that google python lib prints a warning for every execution about a feature that won't be used (and is not necessary to function), which floods the stackdriver logs.
`WARNING: file_cache is unavailable when using oauth2client >= 4.0.0`

I was also reading this blog post: https://www.electricmonk.nl/log/2017/08/06/understanding-pythons-logging-module/
Which says it is better to use a child logger then the root logger to have more control about the logs.

oalbrigt commented:
Thanks. I'll see if I can find a workaround for that.

oalbrigt commented:
You can suppress those Warnings with the following line:
logging.getLogger("googleapiclient.discovery_cache").setLevel(logging.ERROR)

helen-fornazier commented:
> You can suppress those Warnings with the following line:
> logging.getLogger("googleapiclient.discovery_cache").setLevel(logging.ERROR)

I don't like this, because it means we are suppressing those messages in the root logger too, so it won't show up when executing with -v or -D

helen-fornazier commented:
> > You can suppress those Warnings with the following line:
> > logging.getLogger("googleapiclient.discovery_cache").setLevel(logging.ERROR)
> 
> I don't like this, because it means we are suppressing those messages in the root logger too, so it won't show up when executing with -v

Never mind, I am find with this, please check my last commit

oalbrigt commented on agents/gce/fence_gce.py:
```diff
@@ -38,50 +56,141 @@ def get_power_status(conn, options):
 	conn = None
 
-	device_opt = ["port", "no_password", "zone", "project"]
+	hostname = platform.node()
```
You should use options["--plug"] instead of this, to avoid incosistency (and users might use different instance name and hostname).

helen-fornazier replied:
hostname is just for logging to stackdriver

oalbrigt replied:
Yeah, but I dont think it's necessary to import platform and define a new variable here when you can just use options["--plug"] on line 214 instead.

helen-fornazier replied:
hostname is only used by stackdriver logs, if I understood correctly, the action `-o list` does't require `--plug`, so if I remove the hostname, the stackdriver won't be initialized correctly if no `--plug` is provided.

oalbrigt replied:
I guess you could skip stackdriver logging for those calls (as local logging should be sufficient when not fencing or checking the status of an instance).

helen-fornazier replied:
agreed, I'll update the PR

oalbrigt commented:
Looks good to me.

By the way. Is the reset-function guaranteed to be hard-reset? Some systems use soft-reset (pretty much like running "reboot" on the node), which isnt good for fencing, so we should only use cycle as default if it's always hard-reset.

helen-fornazier commented:
It is, from https://cloud.google.com/compute/docs/instances/restarting-an-instance

    Performing a reset on your instance is similar to pressing the reset button on your computer

oalbrigt commented:
Great.

oalbrigt commented on agents/gce/fence_gce.py:
```diff
@@ -97,6 +206,25 @@ def main():
+			handler.setFormatter(formatter)
+			root_logger = logging.getLogger()
+			if options.get('--verbose') is None:
+				root_logger.setLevel(logging.INFO)
```
This doesnt make any sense. The logic for the root_logger is taken care of in fencing.py.

helen-fornazier replied:
yes, but if we use `--stackdriver-logging`, the logs are not being sent to stackdriver because the logging level defaults to WARN and not INFO.

oalbrigt replied:
Cant you just add the handler above (which has the correct level for stackdriver logging) like we do in the fencing library?

Example:
https://github.com/ClusterLabs/fence-agents/blob/master/lib/fencing.py.py#L723

helen-fornazier replied:
I am not sure I understand what you mean, I added the handler here: https://github.com/ClusterLabs/fence-agents/pull/211/files/9ae0a072424fa982e1d18a2cb661628c38601c3a#diff-4a0f3ce2f5196c812280cc52f7aa42b6R222

oalbrigt replied:
Oh. I didnt find that. But doesnt that log correctly based on the handler.setLevel() above?

helen-fornazier replied:
no, this is just a filter to the handler. The root logger get the logs based in the root logger level, then it sends the logs to the handlers which filter the messages again with its own level, but if the logs does't get thought the root logger in the first place, then nothing is sent to the handlers.

oalbrigt replied:
Ah. I guess we'll just have to live with that, and I might improve that later by lowering the other sub-loggers in this case.

helen-fornazier replied:
This is a nice chart flow explaining that: https://docs.python.org/3/howto/logging.html#logging-flow

oalbrigt replied:
Thanks.

oalbrigt commented on agents/gce/fence_gce.py:
```diff
@@ -97,6 +206,25 @@ def main():
+			root_logger = logging.getLogger()
+			if options.get('--verbose') is None:
+				root_logger.setLevel(logging.INFO)
+				logging.getLogger("googleapiclient").setLevel(logging.ERROR)
```
This should be put outside stackdriver-logging, as we dont want to see these warnings for every run.

helen-fornazier replied:
ok, I'll update the PR.

oalbrigt commented:
@helen-fornazier Pacemaker already does that for us, so that's not necessary (telling which fence device is used in the log before the messages).

By the way it's STONITH, not STONISH.

helen-fornazier commented:
@oalbrigt I the changes we agreed on in this last commit, could you please check now? I just wans't sure what to do about setting the level of the logging, so I left it as it was, please check.

Thanks for reviewing this.

oalbrigt commented:
Looks good to me now. And I think the logging is good too.

oalbrigt commented:
Thanks.
